Source/JavaScriptCore/yarr/YarrJIT.cpp

```diff
@@ -1993,252 +1994,373 @@ class YarrGenerator final : public YarrJITInfo {
+        unsigned lastPosition = firstPosition;
+        auto firstChar = op.m_term->patternCharacter;
+        bool have16BitCharacter = !isLatin1(firstChar);
 
         // For case-insesitive compares, non-ascii characters that have different
         // upper & lower case representations are converted to a character class.
-        ASSERT(!op.m_term->ignoreCase() || isASCIIAlpha(ch) || isCanonicallyUnique(ch, m_canonicalMode));
+        ASSERT(!op.m_term->ignoreCase() || isASCIIAlpha(firstChar) || isCanonicallyUnique(firstChar, m_canonicalMode));
 
-        if (op.m_term->ignoreCase() && isASCIIAlpha(ch)) {
+        if (have16BitCharacter && m_charSize == CharSize::Char16 && !U_IS_BMP(firstChar)) {
+            // The first term we are considering is a non-BMP char in a 16 bit pattern. Just try matching it and be done.
 #if CPU(BIG_ENDIAN)
-            ignoreCaseMask |= 32 << (m_charSize == CharSize::Char8 ? 24 : 16);
+            uint64_t charToMatch = firstChar << (m_charSize == CharSize::Char8 ? 24 : 16);
 #else
-            ignoreCaseMask |= 32;
+            uint64_t charToMatch = firstChar;
 #endif
+
+            auto offset = op.m_checkedOffset - op.m_term->inputPosition;
+
+            if (!matchTargets.hasSucceedTarget() || m_ops[opIndex + 1].m_op == YarrOpCode::Term)
+                defaultMatchTargets.appendFailed(jumpIfCharNotEquals(charToMatch, offset, character, op.m_term->ignoreCase()));
+            else
+                matchTargets.appendSucceeded(jumpIfCharEquals(charToMatch, offset, character, op.m_term->ignoreCase()));
+
+            return;
         }
 
-        for (numberCharacters = 1; numberCharacters < maxCharactersAtOnce && nextOp->m_op == YarrOpCode::Term; ++numberCharacters, nextOp = &m_ops[opIndex + numberCharacters]) {
-            PatternTerm* nextTerm = nextOp->m_term;
+        opList.append(&m_ops[opIndex]);
+
+        for (size_t i = opIndex + 1; i < m_ops.size(); ++i) {
+            YarrOp* currOp = &m_ops[i];
+            if (currOp->m_op != YarrOpCode::Term)
+                break;
+
+            PatternTerm* currTerm = currOp->m_term;
 
             // YarrJIT handles decoded surrogate pair as one character if unicode flag is enabled.
             // Note that the numberCharacters become 1 while the width of the pattern character becomes 32bit in this case.
-            if (nextTerm->type != PatternTerm::Type::PatternCharacter
-                || nextTerm->quantityType != QuantifierType::FixedCount
-                || nextTerm->quantityMaxCount != 1
-                || nextTerm->inputPosition != (startTermPosition + numberCharacters)
-                || (U16_LENGTH(nextTerm->patternCharacter) != 1 && m_decodeSurrogatePairs))
+            if (currTerm->quantityType != QuantifierType::FixedCount
+                || currTerm->quantityMaxCount != 1
+                || (currTerm->type != PatternTerm::Type::PatternCharacter
+                    && currTerm->type != PatternTerm::Type::CharacterClass)
```

Comment:
OK, now it is extended for character class.

Reply:
We treat these character classes as holes that we mask around when processing the surrounding pattern characters.